module-system/module-implementations/sov-nft-module/src/call.rs

```diff
@@ -0,0 +1,300 @@
+    ) -> Result<CallResponse> {
+        let (collection_address, _) =
+            self.get_collection_by_name(collection_name, context, working_set)?;
+        let token_identifier = NftIdentifier(token_id, collection_address);
+        let mut nft = self.get_nft_by_id(&token_identifier, working_set)?;
+        nft.exit_if_frozen()?;
+        if let Some(val) = frozen {
+            nft.frozen = val
+        };
+        if let Some(murl) = token_uri {
+            nft.token_uri = murl
+        };
+        self.nfts.set(&token_identifier, &nft, working_set);
+
+        Ok(CallResponse::default())
+    }
+
+    fn exit_if_collection_exists(
+        &self,
+        collection_name: &str,
+        context: &C,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<()> {
+        let creator = context.sender();
+        let ca = get_collection_address(collection_name, creator.as_ref());
+        ensure!(
+            self.collections.get(&ca, working_set).is_none(),
+            format!(
+                "Collection with name: {} already exists creator {}",
+                collection_name, creator
+            )
+        );
+        Ok(())
+    }
+
+    fn exit_if_nft_exists(
+        &self,
+        token_id: TokenId,
+        collection_address: &CollectionAddress<C>,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<()> {
+        let msg = format!(
+            "Collection with address {} already exists",
+            collection_address.0
+        );
+        ensure!(
+            self.nfts
+                .get(
+                    &NftIdentifier(token_id, collection_address.clone()),
+                    working_set
+                )
+                .is_none(),
+            msg
+        );
+        Ok(())
+    }
+
+    fn get_collection_by_address(
+        &self,
+        collection_address: &CollectionAddress<C>,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<Collection<C>> {
+        let c = self.collections.get(collection_address, working_set);
+        if let Some(collection) = c {
+            Ok(collection)
+        } else {
+            Err(anyhow!(
+                "Collection with address: {} does not exist",
+                collection_address.0
+            ))
+        }
+    }
+
+    fn get_collection_by_name(
+        &self,
+        collection_name: &str,
+        context: &C,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<(CollectionAddress<C>, Collection<C>)> {
+        let creator = context.sender();
+        let ca = get_collection_address(collection_name, creator.as_ref());
+        let c = self.collections.get(&ca, working_set);
+        if let Some(collection) = c {
+            Ok((ca, collection))
+        } else {
+            Err(anyhow!(
+                "Collection with name: {} does not exist for creator {}",
+                collection_name,
+                creator
+            ))
+        }
+    }
+
+    fn get_nft_by_id(
```

Comment:
Same question here - should we try to bake the `is_owner` check into this method using the type system. Maybe we have two methods: `get_owned_nft(context, collection_address, nft_id)` which returns an `OwnedfNft` that has a `pub owner` field and only allows `get` access to all other fields, and `get_nft_for_admin`, which has `pub uri` and `pub frozen` fields, but only provides `get` access to the rest of the fields. 

Then, we just need to implement `From<Nft>` and `into<Nft>` for each of these newtypes, and we get full type safety

Reply:
We also have functionality using the `EncodeLike` trait to let these kinds of newtypes be stored directly in the JMT instead of using `From` and `Into`